### ISSUE TITLE
v5 new component: MobileNavMenu

### ISSUE THREAD

**Issue by ss-dimasm** [reapit/elements#236]
**Abstract**

Now that the `MobileNavItem` component is live in the primary branch, we can focus on developing the `MobileNavMenu` component. This ticket outlines the approach we'll take

![{C3C51760-7848-4A09-803E-ADD029AC92B6}](https://github.com/user-attachments/assets/f37216b6-8d9f-47fc-bdee-74048fa786ee)

**Specification**

- Is new component or existing? new
- Is the component being re-named? no
- Is the component design signed off by product and design? yes
- [Guide to component structure & documentation in storybook](https://www.figma.com/design/XJ6qcAV8gHscsUodqJMNEF/Reapit-Elements-production-ready-components?node-id=18-6292&node-type=instance&m=dev)
- [Accessibility requirements](https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/)
- [Link to Design Guidelines](https://www.figma.com/design/pJ96bxoEFStjjIbAPy0Jaj/Design-Guidelines?node-id=1661-8748&node-type=instance&m=dev)

**Developer Checklist**

- [ ] Styles alignment between Design System and Elements
- [ ] Check design tokens in Figma and implement CSS variable tokens if available for relevant component
- [ ] Align with accessibility standards / spec as per above
- [ ] If relevant, break down component into `Styles Only` and `React` component structures
- [ ] Ensure all variants of components are documented as appropriate 
- [ ] Ensure unit test coverage is adequate for component
- [ ] Update documentation in MDX file as per guidelines
- [ ] Changelog updated to reflect a single beta version per component ideally

**Release Checklist**

- [ ] Approved PR merged to main
- [ ] Design & product review and feedback addressed by developer
- [ ] Beta release by product / engineering lead to next beta version

**Additional Context or Information**

**Comment by ss-dimasm** [comment]
Hi @kurtdoherty and @adrian-reapit, I’ll gather some initial information on how this component should be structured. Creating an early POC is necessary, and I’ll return to this conversation once it’s complete, thanks 👍

**Comment by kurtdoherty** [comment]
@ss-dimasm I don't think we're really in a position to start this work yet. The `MobileNavMenu` will need to use new `Drawer` components which aren't implemented yet. Further, I'm probably best positioned to implement these Drawer components as I have a lot of context from recent work done with them in the Console Cloud product.

**Comment by ss-dimasm** [comment]
yes, while am looking at it yesterday this component requires 2 dependencies that need to be solved: (1) Drawer and (2) App switcher; I assume the second dependency is not required at least for now

for addition, the Drawer would have 2 different transitions based on users screen width, on tablet it would have slides left animation and for desktop it would have fade-in and fade-out; I really hope if this also handled in the new `Drawer`

**Comment by kurtdoherty** [comment]
> the Drawer would have 2 different transitions based on users screen width, on tablet it would have slides left animation and for desktop it would have fade-in and fade-out; I really hope if this also handled in the new Drawer

Ooh, I had not seen that part of the mobile nav's spec. I think that may mean it can't/shouldn't depend on Drawer 🤔 ... so you probably can start work on this. 

There'll be a conversation needed around the mobile nav menu trigger's animation. Where Andrei says they can provide a JSON file for the animation, I think that's referring to a [LottieFile](https://lottiefiles.com/). If so, we'll need to decide if that's appropriate given Elements' current "zero-dependency" mandate.

**Comment by ss-dimasm** [comment]
hey @kurtdoherty , I've done with the POC for `MobileNavMenu` and based on the conversation with @adrian-reapit we could drop some of the animation for close button on there

you might check the POC on `poc/mobile-navigation-menu` branch ([files changed](https://github.com/reapit/elements/commit/48f4a551bfb3c03f74fe9f257ecdfd353a480b98)); I also attach the videos on how they would looks (see below)

### Interfaces
about the API, am trying to achieve like how the future `BottomBar` and future `SideBar` structured, however I also have some uncertain condition which I would ask in detail section

```tsx
<MobileNavMenu isOpen={isOpen} onClose={handleOnClose}>
  <MobileNavMenu.Header onClose={handleOnClose} />
  <MobileNavMenu.ItemGroup>
    <MobileNavMenu.Item label="Main Nav Item 1" hasBadge>
      <MobileNavMenu.Item label="Label" href="#item-1.1" />
      <MobileNavMenu.Item label="Label" href="#item-1.2" />
      <MobileNavMenu.Item label="Label" href="#item-1.3" />
      <MobileNavMenu.Item label="Label" href="#item-1.4" />
      <MobileNavMenu.Item label="Label" href="#item-1.5" />
    </MobileNavMenu.Item>
    <MobileNavMenu.Item label="Main Nav Item 2" href="#item-2" />
    <MobileNavMenu.Item label="Main Nav Item 3" href="#item-3" />
    <MobileNavMenu.Item label="Main Nav Item 4" href="#item-4" />
    <MobileNavMenu.Item label="Main Nav Item 5" href="#item-5" />
  </MobileNavMenu.ItemGroup>
  <MobileNavMenu.ItemGroup>
    <MobileNavMenu.Item label="Secondary Nav Item 1" hasBadge isActive>
      <MobileNavMenu.Item label="Label" href="#item-1.1" />
      <MobileNavMenu.Item label="Label" href="#item-1.2" />
      <MobileNavMenu.Item label="Label" href="#item-1.3" isActive />
      <MobileNavMenu.Item label="Label" href="#item-1.4" hasBadge />
      <MobileNavMenu.Item label="Label" href="#item-1.5" />
    </MobileNavMenu.Item>
    <MobileNavMenu.Item label="Secondary Nav Item 2" href="#item-2" />
    <MobileNavMenu.Item label="Secondary Nav Item 3" href="#item-3" />
    <MobileNavMenu.Item label="Secondary Nav Item 4" href="#item-4" />
    <MobileNavMenu.Item label="Secondary Nav Item 5" href="#item-5" />
  </MobileNavMenu.ItemGroup>
  <MobileNavMenu.ItemGroup>
    <MobileNavMenu.Item label="User Item 1" href="#item-1.1" />
    <MobileNavMenu.Item label="User Item 2" href="#item-1.1" />
    <MobileNavMenu.Item label="User Item 3" href="#item-1.1" />
    <MobileNavMenu.Item label="User Item 4" href="#item-1.1" />
    <MobileNavMenu.Item label="User Item 5" href="#item-1.1" />
  </MobileNavMenu.ItemGroup>
</MobileNavMenu>
```

### Details
1. MobileNavMenu
This component will have a role as a parent or starting point from `MobileNavMenu` atoms, and this will scaffolded with native html dialog node. So in underground it would utilize the `useToggleDialogVisibilityEffect` that already have the responsibility to handle the open and close states from a native html dialog

2. MobileNavMenu.Header
Place where the header and their content should belong

**question:** should we also need to bring the `handleClose` with render prop pattern? and not trying to encapsulate the `close` button? I believe it would make user more flexible but it doesn't mean we're following the design spec 

4. MobileNavMenu.ItemGroup
Container of `MobileNavMenu.Item` and it's backed with `ul` native html, this would act like a divider between group of items

5. MobileNavMenu.Item
An alias from `MobileNavItem` that basically is wrapped with `li`

### Preview

and a little attachment on how the POC would be looks like:

**- Mobile (Fade animation)**

https://github.com/user-attachments/assets/c3bec76b-21f3-4f2f-bf2e-eedc03c1eb99

**- Tablet (Slide animation)**

https://github.com/user-attachments/assets/26bcebf2-a856-4dde-9866-0079ded2eff0

**Comment by kurtdoherty** [comment]
@ss-dimasm generally speaking the above all looks pretty good, but I'm concerned about the absence of the open/close icon's animation.

That animation adds delight to the experience for our users and it pretty much expected by them given it's prevalence in other products; I can't see design being satisfied with that animation being descoped. Further, the implementation of that animation will impact the interface design of our MobileNavMenu, so avoiding it now could mean the interface you've described above proves inadequate when the animation is then added in future.

I think Design needs to weigh in here and indicate whether it is appropriate to descope or not. My opinion would be that it is not.

**Comment by ss-dimasm** [comment]
thanks @kurtdoherty, I'll ask in Figma about this concern

**Comment by ss-dimasm** [comment]
> the animation is not critical. It can be added later. We can use a LottieFiles Figma plugin to export Figma animations as Lottie JSON. But I think you need to embed a Lottie player somewhere in the code. Some example of Lottie JSON animations here: https://lottiefiles.com/free-animations/json
We may also find a way to export the animation as an animated SVG. It's probably best if you build the animation with CSS to avoid any complications.

From Design team perspective from the raised query, it's not critical and we could add in later stage. Personal opinion, the button animation need to have their own ticket since it's need a special attention (different animations that we've did)

**Comment by kurtdoherty** [comment]
Right, but it sounds like they expect it to be implemented (now or later is just a matter of the level of quality we want to deliver now vs later). Either way, it means we need to take care to implement the hamburger button in a manner that allows us to easily animate it in future. Your current API does not seem to allow that as the close button inside the menu is not the same as the button that opens in the menu.

**Comment by ss-dimasm** [comment]
yes, it should be a single node that placed between `TopBar` and `MobileNavMenu`

the `TopBar` has expose the `MobileNav` compound component which it would be a place where the actual `MobileNavMenu` implementation is live (example https://github.com/reapit/elements/blob/main/src/components/top-bar/top-bar.stories.tsx#L94-L96)

```tsx
<TopBar.MobileNav>
  <MobileNavMenu isOpen={isOpen} onClose={handleOnClose}>
    <MobileNavMenu.Header>
      <MobileNavMenu.ToggleButton isOpen={isOpen} onOpen={handleOnOpen} onClose={handleOnClose} />
    </MobileNavMenu.Header>
    // ... rest mobilenavmenu implementation
  </MobileNavMenu>
</TopBar.MobileNav>
```

I think we could add additional encapsulation from `NavIconItem` or just have a new fresh component (`MobileNavMenu.ToggleButton`) that have a responsibility to open and close the mobile nav menu including with the animation

so at the end, we only have one actual node and it's not isolated in the `MobileNavMenu.Header` and the entire API still persist

**Comment by kurtdoherty** [comment]
**thought/question:** Do you think consumers need to be able to control the open state of the mobile nav menu? I'm struggling to think of a scenario where a consumer would need to be the one that determines when the mobile nav menu is open or not, so I wonder if that should just be internal to the MobileNavMenu?

**Comment by ss-dimasm** [comment]
> so I wonder if that should just be internal to the MobileNavMenu

I believe the controller should be placed outside the MobileNavMenu, but not within the AppBar, making it more feasible to cover scenarios where the MobileNavMenu is in a controlled state with specific nodes (i.e. more than the ToggleButton)

**Comment by ss-dimasm** [comment]
> What kind of scenarios are you thinking of here? I can't think of any reason why a consumer would need or want to control the nav menu's open state...similar to our Menu component...

Delivering the flexibility, and be a portable component. It would make easier for the `AppBar` (not `Topbar`) component to integrate the open or closed states of `MobileNavMenu`. Additionally, I think consumers could utilize the state value or the callbacks to implement tracking, monitoring clicks on the entry point or state changes

**Comment by kurtdoherty** [comment]
If MobileNavMenu handles the open state internally, AppBar has _less_ work to do. Also, remember that there's no agreement around the need for any kind of all-in-one AppBar component, so we should be careful not to use that as our justification for making certain decisions with other components.

Regarding flexibility: I don't understand why we need it here? We would be placing extra work on the consumer. When _won't_ a consumer want the mobile nav menu to open when the trigger is clicked?

Regarding analytics tracking: consumers don't need to be able to control the `isOpen` state in order to track click events. It's enough to just be able to provide a custom `onClick` handler to the various atoms involved in the mobile nav menu, so we already have the appropriate level of flexibility here.

Lastly, if we were to discover a need for a consumer to explicitly control the `isOpen` state in future, we'd be able to facilitate that even if the mobile nav menu is, by default, handling that state internally.

**Comment by ss-dimasm** [comment]
I appreciate your concerns about unnecessary consumer flexibility. Your points regarding analytics tracking and potential future enhancements are well-taken. For now, handling the open state internally within MobileNavMenu seems reasonable

Is there anything left for the `MobileNavMenu`?

**Comment by kurtdoherty** [comment]
One issue that popped up in the SideBar discussion in #154 is some complexity around the handling of `isActive` for menu items. I think you and @nurm717123 could put your heads together to explore that in more detail via proper integration with RR6. I don't think what we have with mobile nav items will actually behave how we need it to when the user navigates around a product. See my second suggestion in https://github.com/reapit/elements/issues/154#issuecomment-2540332383 for more conversation around this.

**Comment by ss-dimasm** [comment]
thanks, a valid point to you've spot since last week 👍; the current `MobileNavItem` that have a children couldn't listen based on the outside of the component because it's have the isolated state; we'll discuss it more further on this

https://github.com/reapit/elements/blob/main/src/components/mobile-nav-item/mobile-nav-item.atoms.tsx#L81

**Comment by ksolanki7** [comment]
@ss-dimasm Please review https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-select-only/ doc for referencing for a11y.

**Comment by kurtdoherty** [comment]
@ksolanki7 Apologies if I'm missing something, but what does combobox a11y have to do with the MobileNavMenu?

**Comment by ksolanki7** [comment]
@kurtdoherty I added it as a reference for keyboard navigation in the MobileNavMenu. Currently, accessibility support is limited to navigation using the `Tab` key. However, full keyboard support for the menu should include navigation with the Up, Down, and Escape keys, among others. That is the reason I added above reference for inspiration.

**Comment by ss-dimasm** [comment]
@kurtdoherty  well, there are 4 PRs before, they all need a sequential review but as it no activity in the meantime, so I am not sure what we should do for this ticket at that time

https://github.com/user-attachments/assets/3d88e8b0-ddaa-4221-bf7c-a2beb0217851

PRs reference (ordered):
1. https://github.com/reapit/elements/pull/397
2. https://github.com/reapit/elements/pull/398
3. https://github.com/reapit/elements/pull/400
4. https://github.com/reapit/elements/pull/401
